File: app/schemas/connection.py
from __future__ import annotations

from datetime import datetime, timezone
from enum import Enum

from .base import SchemaModel
from .pydantic_compat import Field, model_validator


class DatabaseType(str, Enum):
    POSTGRESQL = "postgresql"
    MYSQL = "mysql"
    SQLITE = "sqlite"


class ConnectionConfig(SchemaModel):
    name: str
    db_type: DatabaseType
    host: str | None = None
    port: int | None = None
    username: str | None = None
    password: str | None = None
    database: str
    file_path: str | None = None

    @model_validator(mode="after")
    def validate_required_fields(self):
        if self.db_type in {DatabaseType.POSTGRESQL, DatabaseType.MYSQL}:
            missing = [
                name
                for name in ("host", "port", "username", "password", "database")
                if getattr(self, name) in (None, "")
            ]
            if missing:
                raise ValueError(f"Missing required fields for {self.db_type.value}: {', '.join(missing)}")
        if self.db_type == DatabaseType.SQLITE and not (self.file_path or self.database):
            raise ValueError("SQLite connections require file_path or database")
        return self


class ConnectionCreate(ConnectionConfig):
    pass


class ColumnInfo(SchemaModel):
    name: str
    type: str
    nullable: bool = True
    default: str | None = None
    comment: str | None = None
    is_primary_key: bool = False
    is_foreign_key: bool = False


class TableSchema(SchemaModel):
    name: str
    comment: str | None = None
    columns: list[ColumnInfo] = Field(default_factory=list)
    description: str | None = None
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class ConnectionStatus(SchemaModel):
    id: str
    name: str
    db_type: DatabaseType
    database: str
    host: str | None = None
    port: int | None = None
    username: str | None = None
    file_path: str | None = None
    is_online: bool = False
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    schema_updated_at: datetime | None = None


class ConnectionTestResult(SchemaModel):
    connection_id: str
    success: bool
    message: str
    latency_ms: float = 0.0
    tested_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))


class SchemaCacheEntry(SchemaModel):
    connection_id: str
    tables: list[TableSchema] = Field(default_factory=list)
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

File: NL2SQL Agent Frontend Development/src/app/types/connection.ts
import type { ConnectionRAGStatus } from './rag';

export type DBType = 'postgresql' | 'mysql' | 'sqlite';

export interface ConnectionConfig {
  name: string;
  db_type: DBType;
  host?: string;
  port?: number;
  username?: string;
  password?: string;
  database: string;
  // SQLite 文件路径
  file_path?: string;
}

export interface Connection extends ConnectionConfig, ConnectionRAGStatus {
  id: string;
  is_online: boolean;
  created_at: string;
  updated_at?: string;
  schema_updated_at?: string | null;
}

export interface ColumnInfo {
  name: string;
  type: string;
  nullable?: boolean;
  default?: string | null;
  comment?: string;
  is_primary_key?: boolean;
  is_foreign_key?: boolean;
  // legacy aliases
  primary_key?: boolean;
  foreign_key?: string;
}

export interface TableSchema {
  name: string;
  comment?: string;
  columns: ColumnInfo[];
  description?: string;
  updated_at?: string;
  // legacy aliases
  table_name?: string;
  table_comment?: string;
}

// Schema 缓存条目
export interface SchemaCacheEntry {
  connection_id: string;
  tables: TableSchema[];
  updated_at: string;
}

export interface ConnectionStatus extends Partial<ConnectionRAGStatus> {
  id: string;
  name?: string;
  db_type?: DBType;
  database?: string;
  host?: string;
  port?: number;
  username?: string;
  file_path?: string;
  is_online?: boolean;
  created_at?: string;
  updated_at?: string;
  schema_updated_at?: string | null;
  status?: 'online' | 'offline';
  message?: string;
}

// 连接测试结果
export interface ConnectionTestResult {
  connection_id?: string;
  success: boolean;
  message: string;
  latency_ms?: number;
  tested_at?: string;
}
